pkg/compact/downsample/downsample.go

```diff
@@ -734,3 +739,104 @@ func SamplesFromTSDBSamples(samples []tsdbutil.Sample) []sample {
 	}
 	return res
 }
+
+// GatherNoDownsampleMarkFilter is a block.Fetcher filter that passes all metas.
+// While doing it, it gathers all no-downsample-mark.json markers.
+type GatherNoDownsampleMarkFilter struct {
+	logger                log.Logger
+	bkt                   objstore.InstrumentedBucketReader
+	noDownsampleMarkedMap map[ulid.ULID]*metadata.NoDownsampleMark
+	concurrency           int
+	mtx                   sync.Mutex
+}
+
+// NewGatherNoDownsampleMarkFilter creates GatherNoDownsampleMarkFilter.
+func NewGatherNoDownsampleMarkFilter(logger log.Logger, bkt objstore.InstrumentedBucketReader) *GatherNoDownsampleMarkFilter {
+	return &GatherNoDownsampleMarkFilter{
+		logger:      logger,
+		bkt:         bkt,
+		concurrency: 1,
+	}
+}
+
+// NoDownsampleMarkedBlocks returns block ids that were marked for no downsample.
+func (f *GatherNoDownsampleMarkFilter) NoDownsampleMarkedBlocks() map[ulid.ULID]*metadata.NoDownsampleMark {
+	f.mtx.Lock()
+	copiedNoDownsampleMarked := make(map[ulid.ULID]*metadata.NoDownsampleMark, len(f.noDownsampleMarkedMap))
+	for k, v := range f.noDownsampleMarkedMap {
+		copiedNoDownsampleMarked[k] = v
+	}
+	f.mtx.Unlock()
+
+	return copiedNoDownsampleMarked
+}
+
+// Filter passes all metas, while gathering no downsample markers.
+func (f *GatherNoDownsampleMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced block.GaugeVec, modified block.GaugeVec) error {
```

Comment:
I wonder if we need to reimplement the filtering logic just because of a new marker. For the short term It should be fine to have it this way, but maybe you can leave a TODO comment to merge the no compact and no downsample filters?

Reply:
Good idea. I'll add in a TODO and open an issue/PR to make that change.